### PR TITLE
Publish to s3

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR upgrades Gradle to 7.1.1 & Android Gradle Plugin to 4.2.2. It also adds support for publishing library artifacts to our S3 repository with a Buildkite job.

I'll be leaving several line comments for details which should be easier to follow & discuss than listing them here.

**To test:**
The changes can be tested from https://github.com/Automattic/simplenote-android/pull/1446.

### CODE REVIEW

File: .buildkite/pipeline.yml
Comment (oguzkocer) [inline .buildkite/pipeline.yml:2]: For now I've only added a step for publishing, but in the long term, I think we'll want to add unit tests, lint and if possible connected tests as a dependency.

File: README.md
Comment (oguzkocer) [inline README.md:5]: Travis doesn't seem to be active and afaik we've not used Travis for a long time, so I removed the config & status for it.

File: README.md
Comment (oguzkocer) [inline README.md:34]: Publishing is now done from CI, no instructions are necessary.

File: Simperium/build.gradle
Comment (oguzkocer) [inline Simperium/build.gradle:1]: I've updated this project to Gradle 7.1.1, Android Gradle Plugin 4.2.2 and to use plugin DSL. We'd have wanted to do this anyway, so I figured I'd included it in this update.

File: Simperium/build.gradle
Comment (oguzkocer) [inline Simperium/build.gradle:84]: This is the only major and somewhat questionable change in this PR. The version name used to publish the library would not be correct, so I don't think we should rely on it at all. I've removed all the relevant code for this and left 2 TODOs for what to do about them.

File: Simperium/build.gradle
Comment (oguzkocer) [inline Simperium/build.gradle:56]: This is just moved from the root build file to remove the `allprojects` block.

File: Simperium/build.gradle
Comment (oguzkocer) [inline Simperium/build.gradle:66]: I picked `com.automattic` as the domain. We can change it to `simperium` and the `artifactId` to `simperium-android` if it's preferred. This decision changes how the dependency will be added to the projects, [like so](https://github.com/Automattic/simplenote-android/pull/1446/files#diff-43f4c0dc2dc319470a77750c75fc377f96ac31e7183e84fac0ee5ada14aa1042R84).

File: Simperium/src/main/java/com/simperium/client/Channel.java
Comment (oguzkocer) [inline Simperium/src/main/java/com/simperium/client/Channel.java]: This is one of the TODOs I mentioned for removing the generated version information. I'd like to get some help on figuring out how to address this from @danilo04 & maybe @nsakaimbo as you'd have more information about its usage than I do.

To summarize, the version information is not available at the time of code compilation so it can not be used within the library. This project was using git hash for these values which we can still use. However, I'd much prefer if the build information is provided by the client as a passed value instead.

[It's worth noting that Android Gradle Plugin removed the version information for libraries from generated the `BuildConfig` class for similar reasons.](https://developer.android.com/studio/releases/gradle-plugin#version_properties_removed_from_buildconfig_class_in_library_projects)

File: Simperium/src/main/java/com/simperium/android/AndroidClient.java
Comment (oguzkocer) [inline Simperium/src/main/java/com/simperium/android/AndroidClient.java]: This value was always `android` so removing the generated `Version` file doesn't have any impact on this.

File: Simperium/src/main/java/com/simperium/Simperium.java
Comment (oguzkocer) [inline Simperium/src/main/java/com/simperium/Simperium.java]: This is another TODO that's affected by removing the generated `Version` file.

File: .buildkite/pipeline.yml
Comment (danilo04) [reply .buildkite/pipeline.yml:2]: AFAIK, simperium runs all tests (including unit tests) in their own suite that runs through connected tests.

File: README.md
Comment (danilo04) [reply README.md:34]: Hi @oguzkocer. When is publishing done specifically? After merge to develop or at what phase?

File: Simperium/build.gradle
Comment (danilo04) [reply Simperium/build.gradle:1]: Great. You save me headaches for sure. Thanks!

File: README.md
Comment (oguzkocer) [reply README.md:34]: @danilo04 Sorry, I should have listed that in the PR description 🤦 

Here is the list of when artifacts will be published and what the version name will be:

* For all tags -> `{tag-name}`
* For all commits in `develop` (so PR merges) -> `develop-{sha1}`
* For all commits for open PRs - you can open a draft PR to get it to publish. -> `{prNumber}-{sha1}`

These settings can be tweaked, but this is the way we do it for all Android libraries.

---

Typical workflow is: 

* Open a PR and test the published version (which will be printed out in the Buildkite job and follows the naming pattern above)
* Merge the PR into `develop`
* Update `simplenote-android` PR to use the merge commit which is `develop-{sha1}`
* Merge the `simplenote-android` PR
* At the time of release, we create a new tag for the `simperium-android` and update the version value in `simplenote-android` to the tagged value.

File: .buildkite/pipeline.yml
Comment (oguzkocer) [reply .buildkite/pipeline.yml:2]: Does that include unit tests for this repo? That's quite interesting.

File: README.md
Comment (danilo04) [reply README.md:34]: Thanks for the explanation. That makes it easier also for testing the changes with `simplenote-android`.

Comment (oguzkocer) [reply README.md:34]: I'll add composite build support which will make it even easier to test it. I just need a little bit of time to finish up some other tasks. 😅

File: Simperium/src/main/java/com/simperium/client/Channel.java
Comment (danilo04) [reply Simperium/src/main/java/com/simperium/client/Channel.java]: Based on what I found by asking the team, this is important information for debugging purposes. It is important to identify what client and version generated the different traces.

File: Simperium/build.gradle
Comment (mokagio) [reply Simperium/build.gradle:66]: FWIW, keeping it under the Automattic reverse domain name make sense to me.

File: Simperium/src/main/java/com/simperium/client/Channel.java
Comment (mokagio) [reply Simperium/src/main/java/com/simperium/client/Channel.java]: I'm sure that if there was a way to access these at runtime you would have already mentioned it.

The only thing I can think of it that, if we need certain identifiers to help observability, we might add custom ones. E.g. we could have a custom field that reads the Git commit SHA1 and exposes it in `BuildConfig` or in a more appropriate location. Just an idea, I don't actually know if it's possible 😅

**Review by mokagio — APPROVED** [review]
I tested it on the companion [Simplenote Android PR](https://github.com/Automattic/simplenote-android/pull/1446) and it worked. The code changes make sense to me, with my limited understanding of Android tooling.

File: README.md
Comment (AliSoftware) [reply README.md:34]: btw just wanted to take the occasion to thank @oguzkocer again on all his work on that, and especially the new way we use for publishing Android libraries. 😍 
With the auto-versioning following the `develop-{sha1}`/`{prNum}-{sha1}` patterns you mentioned above, which means that developers don't need to create "beta tags" on the libs anymore, this make testing lib changes way easier for everybody! ❤️ 

_(The only tricky part remaining is for the release manager to find the right "next version number" to use at release time i.e. to decide if they need to bump major vs minor vs patch, but usually minor bump makes sense for Android libs given we don't rely on SemVer in Gradle for those anyway, so that's not a big deal anyway)_

Comment (AliSoftware) [reply README.md:34]: Even if the manual instructions are not necessary anymore, do you think I'd still make sense to mention in the readme the auto-publishing by CI and the version name patterns we use (i.e. put your great explanations from [above](https://github.com/Simperium/simperium-android/pull/232/files#r690695360) in the README)?

**Review by AliSoftware — APPROVED** [review]
All the changes make sense to me, at least to my knowledge of the migration to publish-to-s3 and Gradle7.

As for the removal of the version constants, I'll leave the Simperium devs chime in and decide.
I think, like Gio said, that if we really indeed to provide them (e.g. to include them in the requests for debugging purposes) we should provide them via `BuildConfig` fields instead so that they match the actual version name used during S3 publishing, instead of being different and inconsistent with the lib's actual version name.

File: README.md
Comment (danilo04) [reply README.md:34]: @AliSoftware As one of the developers, I would say yes, because a new developer can start working immediately with this information.

File: .buildkite/pipeline.yml
Comment (danilo04) [reply .buildkite/pipeline.yml:2]: Yes, I think the connected tests works as the test suite that loads the unit tests.

File: Simperium/src/main/java/com/simperium/client/Channel.java
Comment (danilo04) [reply Simperium/src/main/java/com/simperium/client/Channel.java]: That would mean some kind of hook in the gradle plugin that sits in between the version definition (reads git commit sha1) and the build. We could then write the `Version` file and then build.

**Comment by oguzkocer** [comment]
@danilo04 @AliSoftware @mokagio @jkmassel @nsakaimbo About the generated version file not being available anymore:

I've reverted the changes in 2fef66a025f4dade3e23fca37de5bddce1d5de19 and made the generated Version file available again. I've then updated the version name to `project.ext.properties.extra_version_name` in 84683943ee6858ac49d110f1039b5ab57e0d329a.

So.. I kind of lied - not intentionally - a little bit when I said we can not access the published version. There is no direct way to access it, that's still true - but internally publish-to-s3 Gradle plugin does add the information to the project properties. I didn't and still don't want to use this information, because it's not intended to be consumed by clients. However, due to the current deadline for merging this PR, I've opted to do this since it's the smallest change we can do to preserve the current behavior. I'd like to be clear that I consider this a hack, at least in its current state, and should be updated once we are not pressed for time.

You can test these changes as such:
* Run `./gradlew build && cat Simperium/build/generated/source/version/com/simperium/Version.java` and verify that the version used is `null`. This is the case for development builds.
* Run `./gradlew prepareToPublishToS3 --tag-name=0.3 build && cat Simperium/build/generated/source/version/com/simperium/Version.java` and verify that the version used is now `0.3`. You can change the value and the command line parameters for different results.

I've intentionally kept the changeset to minimum as there are some Gradle optimizations to be done in the reverted code. I'm hoping that we'll get rid of the code completely soon.

@danilo04 Could you test the [updated Simplenote-Android PR](https://github.com/Automattic/simplenote-android/pull/1446) to verify that version information is sent to backend correctly?

**Comment by danilo04** [comment]
Hi @oguzkocer, everything is working correctly, the only thing that is unknown is the `build` because the `gitHash()` is failing. These are the values that are compiled:

```java
extra.put("build", "<unknown>");
extra.put("version", "232-84683943ee6858ac49d110f1039b5ab57e0d329a");
extra.put("client", "android-232-84683943ee6858ac49d110f1039b5ab57e0d329a");
```

**Review by danilo04 — APPROVED** [review]
I would say with `version` and `client` (since they contain the specific commit hash), we are good to go for now. I will approve this PR. 

@oguzkocer if there is nothing to add to this PR, then we can merge. Thank you for hard work on this. We can create the issues to add improvements in future PRs.

**Comment by oguzkocer** [comment]
@danilo04 The `build` value [that's sent](https://github.com/Simperium/simperium-android/blob/develop/Simperium/src/main/java/com/simperium/client/Channel.java#L543) is [not changed](https://github.com/Simperium/simperium-android/blob/publish-to-s3/Simperium/build.gradle#L101). I imagine this value is sent as `<unknown>` in current `develop` as well - unless I am missing something - so, it shouldn't be related to this PR 🤞

**Comment by mokagio** [comment]
Nice work @oguzkocer! 👏

**Comment by danilo04** [comment]
@oguzkocer I will add publish notes to the `README.md` using your explanation in (https://github.com/Simperium/simperium-android/pull/232#discussion_r690695360) in a working PR I have.

**Pull request merged**
